File: Packages/NoopPush/Sources/NoopPush/PushCapabilities.swift
import Foundation

public struct PushCapabilities: Sendable {
    public let appendTables: Set<PushAppendTable>
    public let mutableTables: Set<PushMutableTable>
    public let binaryTables: Set<PushBinaryTable>
    public let protocolVersion: String
    public let receiverStateId: String
    /// Server-resolved enrollment identity. Generic protocol fixtures may omit these, but enrolled
    /// clients must pin both before opening local storage or uploading records.
    public let userId: String?
    public let sourceId: String?
    /// Direct-to-bucket lane advertised at protocol 1.2 and later. `nil` disables binary upload:
    /// raw rows stay local rather than posting inline into a `use_object_lane` refusal.
    public let objectLane: PushObjectLane?

    public var isEmpty: Bool { appendTables.isEmpty && mutableTables.isEmpty && binaryTables.isEmpty }

    public var wireNames: [String] {
        PushAppendTable.allCases.filter { appendTables.contains($0) }.map(\.wireName)
            + PushMutableTable.allCases.filter { mutableTables.contains($0) }.map(\.wireName)
            + PushBinaryTable.allCases.filter { binaryTables.contains($0) }.map(\.wireName)
    }

    public static let unscopedReceiverStateId = "00000000-0000-4000-8000-000000000000"

    public static let all = PushCapabilities(
        appendTables: Set(PushAppendTable.allCases.filter { !$0.isScalarExtension }),
        mutableTables: Set(PushMutableTable.allCases),
        binaryTables: Set(PushBinaryTable.allCases)
    )

    public init(
        appendTables: Set<PushAppendTable>,
        mutableTables: Set<PushMutableTable>,
        binaryTables: Set<PushBinaryTable> = [],
        protocolVersion: String = PushProtocol.version,
        receiverStateId: String = unscopedReceiverStateId,
        objectLane: PushObjectLane? = nil,
        userId: String? = nil,
        sourceId: String? = nil
    ) {
        self.appendTables = appendTables
        self.mutableTables = mutableTables
        self.binaryTables = binaryTables
        self.protocolVersion = protocolVersion
        self.receiverStateId = receiverStateId
        self.objectLane = objectLane
        self.userId = userId
        self.sourceId = sourceId
    }

    public static func parse(_ bytes: Data) throws -> PushCapabilities {
        guard bytes.count <= PushProtocolLimits.maxAckBytes else {
            throw PushProtocolException("capabilities exceed size limit")
        }
        guard let obj = try JSONSerialization.jsonObject(with: bytes) as? [String: Any] else {
            throw PushProtocolException("capabilities are not valid JSON")
        }
        let required = Set(["type", "protocolVersion", "receiverStateId", "streams"])
        let actual = Set(obj.keys)
        guard required.isSubset(of: actual) else {
            throw PushProtocolException("capabilities are missing required protocol 1.0 members")
        }
        if actual.contains(where: { PushProtocol.forbiddenRemoteControlMembers.contains($0) }) {
            throw PushProtocolException("capabilities contain forbidden remote-control metadata")
        }
        guard obj["type"] as? String == "capabilities" else {
            throw PushProtocolException("unsupported capability document")
        }
        let version = obj["protocolVersion"] as? String ?? ""
        guard version == PushProtocol.version || version == PushProtocol.binaryVersion || PushProtocol.isObjectVersion(version) else {
            throw PushProtocolException("unsupported capability document")
        }
        guard let receiverStateId = obj["receiverStateId"] as? String, isCanonicalUuid(receiverStateId) else {
            throw PushProtocolException("capabilities.receiverStateId must be a canonical UUID")
        }
        let userId = try optionalCanonicalUuid(obj, key: "userId")
        let sourceId = try optionalCanonicalUuid(obj, key: "sourceId")
        guard let streams = obj["streams"] as? [Any] else {
            throw PushProtocolException("capabilities.streams must be an array")
        }
        let appendByName = Dictionary(uniqueKeysWithValues: PushAppendTable.allCases.map { ($0.wireName, $0) })
        let mutableByName = Dictionary(uniqueKeysWithValues: PushMutableTable.allCases.map { ($0.wireName, $0) })
        let binaryByName = Dictionary(uniqueKeysWithValues: PushBinaryTable.allCases.map { ($0.wireName, $0) })
        var seen = Set<String>()
        var append = Set<PushAppendTable>()
        var mutable = Set<PushMutableTable>()
        var binary = Set<PushBinaryTable>()
        for item in streams {
            guard let name = item as? String else {
                throw PushProtocolException("capability stream names must be strings")
            }
            guard seen.insert(name).inserted else {
                throw PushProtocolException("duplicate capability stream")
            }
            if let table = appendByName[name] {
                if !table.isScalarExtension || version != PushProtocol.version { append.insert(table) }
            } else if let table = mutableByName[name] {
                mutable.insert(table)
            } else if let table = binaryByName[name] {
                binary.insert(table)
            }
        }
        // A malformed object-lane block disables the lane (rows are
        // retained) rather than failing the whole capability negotiation.
        let objectLane: PushObjectLane?
        if PushProtocol.isObjectVersion(version), let laneObject = obj["objectLane"] as? [String: Any] {
            objectLane = parseObjectLane(laneObject, binaryByName: binaryByName)
        } else {
            objectLane = nil
        }
        return PushCapabilities(
            appendTables: append,
            mutableTables: mutable,
            binaryTables: binary,
            protocolVersion: version,
            receiverStateId: receiverStateId,
            objectLane: objectLane,
            userId: userId,
            sourceId: sourceId
        )
    }

    private static func parseObjectLane(
        _ obj: [String: Any],
        binaryByName: [String: PushBinaryTable]
    ) -> PushObjectLane? {
        guard let endpoint = obj["endpoint"] as? String,
              endpoint.count <= 256,
              endpoint.hasPrefix("/"), !endpoint.hasPrefix("//"),
              endpoint.range(of: #"\s"#, options: .regularExpression) == nil else { return nil }
        guard let maxObjectBytes = jsonInt64(obj["maxObjectBytes"]),
              maxObjectBytes > 0, maxObjectBytes <= Int64(PushProtocolLimits.maxObjectWireBytes) else { return nil }
        var urlTtlSec: Int64? = nil
        if let ttl = jsonInt64(obj["urlTtlSec"]) {
            guard ttl > 0 else { return nil }
            urlTtlSec = ttl
        }
        guard let streamItems = obj["streams"] as? [Any] else { return nil }
        var laneStreams = Set<PushBinaryTable>()
        for item in streamItems {
            guard let name = item as? String else { return nil }
            guard let table = binaryByName[name] else { continue } // unknown stream names are ignored
            guard laneStreams.insert(table).inserted else { return nil }
        }
        guard !laneStreams.isEmpty else { return nil }
        let completionMode: PushObjectCompletionMode?
        if let modes = obj["completionModes"] as? [String], modes.count <= 2,
           Set(modes).count == modes.count, modes.allSatisfy({ ["sync", "async-v1"].contains($0) }),
           modes.contains("async-v1") { completionMode = .asynchronousV1 }
        else { completionMode = nil }
        return PushObjectLane(
            endpoint: endpoint,
            maxObjectBytes: maxObjectBytes,
            urlTtlSec: urlTtlSec,
            streams: laneStreams,
            completionMode: completionMode
        )
    }

    /// JSONSerialization bridges JSON booleans to NSNumber as well; a lane cap encoded as `true`
    /// is malformed, not 1.
    private static func jsonInt64(_ value: Any?) -> Int64? {
        guard let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID() else { return nil }
        let double = number.doubleValue
        guard double.rounded() == double, double >= Double(Int64.min), double <= Double(Int64.max) else { return nil }
        return number.int64Value
    }

    private static func isCanonicalUuid(_ value: String) -> Bool {
        guard let uuid = UUID(uuidString: value) else { return false }
        return uuid.uuidString.lowercased() == value
    }

    private static func optionalCanonicalUuid(_ object: [String: Any], key: String) throws -> String? {
        guard let raw = object[key] else { return nil }
        guard let value = raw as? String, isCanonicalUuid(value) else {
            throw PushProtocolException("capabilities.\(key) must be a canonical UUID")
        }
        return value
    }
}

public enum PushCapabilitiesResult: Sendable {
    case available(PushCapabilities)
    case rejected(reason: String, retryable: Bool, failure: PushFailure?)
}

extension PushTransport {
    public func capabilities() async throws -> PushCapabilitiesResult {
        .available(.all)
    }

    public func postBinary(_ batch: PushBinaryBatch) async throws -> PushTransportResponse {
        throw PushTransportException(PushFailure(code: .localData))
    }
}

File: android/app/src/main/java/com/noop/push/PushCapabilities.kt
package com.noop.push

import org.json.JSONArray
import org.json.JSONObject

/** A receiver may only narrow the fixed protocol registry, never name new data. */
data class PushCapabilities(
    val appendTables: Set<PushAppendTable>,
    val mutableTables: Set<PushMutableTable>,
    val binaryTables: Set<PushBinaryTable> = emptySet(),
    val protocolVersion: String = PushProtocol.VERSION,
    val receiverStateId: String = UNSCOPED_RECEIVER_STATE_ID,
    val objectLane: PushObjectLane? = null,
) {
    val isEmpty: Boolean get() = appendTables.isEmpty() && mutableTables.isEmpty() && binaryTables.isEmpty()
    val wireNames: List<String> get() =
        PushAppendTable.entries.filter { it in appendTables }.map { it.wireName } +
            PushMutableTable.entries.filter { it in mutableTables }.map { it.wireName } +
            PushBinaryTable.entries.filter { it in binaryTables }.map { it.wireName }

    companion object {
        /** Test/local default. Network capability documents must always provide a real receiver state id. */
        const val UNSCOPED_RECEIVER_STATE_ID = "00000000-0000-4000-8000-000000000000"
        val ALL = PushCapabilities(
            PushAppendTable.entries.toSet(),
            PushMutableTable.entries.toSet(),
            PushBinaryTable.entries.toSet(),
        )

        fun parse(bytes: ByteArray): PushCapabilities {
            if (bytes.size > PushProtocol.MAX_ACK_BYTES) {
                throw PushProtocolException("capabilities exceed size limit")
            }
            val obj = try {
                JSONObject(bytes.toString(Charsets.UTF_8))
            } catch (_: Throwable) {
                throw PushProtocolException("capabilities are not valid JSON")
            }
            val required = setOf("type", "protocolVersion", "receiverStateId", "streams")
            val actualMembers = obj.keys().asSequence().toSet()
            if (!actualMembers.containsAll(required)) {
                throw PushProtocolException("capabilities are missing required protocol 1.0 members")
            }
            if (actualMembers.any { it in PushProtocol.FORBIDDEN_REMOTE_CONTROL_MEMBERS }) {
                throw PushProtocolException("capabilities contain forbidden remote-control metadata")
            }
            val version = obj.opt("protocolVersion") as? String ?: ""
            if (obj.opt("type") != "capabilities" ||
                version !in setOf(PushProtocol.VERSION, "1.1", PushProtocol.OBJECT_VERSION, "1.3", "1.4")
            ) {
                throw PushProtocolException("unsupported capability document")
            }
            val receiverStateId = (obj.opt("receiverStateId") as? String)?.takeIf(::isCanonicalUuid)
                ?: throw PushProtocolException("capabilities.receiverStateId must be a canonical UUID")
            val array = obj.opt("streams") as? JSONArray
                ?: throw PushProtocolException("capabilities.streams must be an array")
            val appendByName = PushAppendTable.entries.associateBy { it.wireName }
            val mutableByName = PushMutableTable.entries.associateBy { it.wireName }
            val binaryByName = PushBinaryTable.entries.associateBy { it.wireName }
            val seen = mutableSetOf<String>()
            val append = linkedSetOf<PushAppendTable>()
            val mutable = linkedSetOf<PushMutableTable>()
            val binary = linkedSetOf<PushBinaryTable>()
            for (index in 0 until array.length()) {
                val name = array.opt(index) as? String
                    ?: throw PushProtocolException("capability stream names must be strings")
                if (!seen.add(name)) throw PushProtocolException("duplicate capability stream")
                appendByName[name]?.let { if (!it.isScalarExtension || version != PushProtocol.VERSION) append += it }
                    ?: mutableByName[name]?.let { mutable += it }
                    ?: binaryByName[name]?.let { binary += it }
            }
            val objectLane = if (version in setOf(PushProtocol.OBJECT_VERSION, "1.3", "1.4") && obj.opt("objectLane") is JSONObject) {
                parseObjectLane(obj.getJSONObject("objectLane"), binaryByName)
            } else {
                null
            }
            return PushCapabilities(append, mutable, binary, version, receiverStateId, objectLane)
        }

        private fun parseObjectLane(
            obj: JSONObject,
            binaryByName: Map<String, PushBinaryTable>,
        ): PushObjectLane? {
            val endpoint = obj.opt("endpoint") as? String ?: return null
            if (endpoint.length > 256 || !endpoint.startsWith("/") || endpoint.startsWith("//") ||
                endpoint.contains(Regex("\\s"))
            ) {
                return null
            }
            val maxObjectBytes = jsonInt64(obj.opt("maxObjectBytes")) ?: return null
            if (maxObjectBytes <= 0L || maxObjectBytes > PushProtocol.MAX_OBJECT_WIRE_BYTES) return null
            val urlTtlSec = when (val ttl = jsonInt64(obj.opt("urlTtlSec"))) {
                null -> null
                else -> if (ttl > 0L) ttl else return null
            }
            val streamItems = obj.opt("streams") as? JSONArray ?: return null
            val laneStreams = linkedSetOf<PushBinaryTable>()
            for (index in 0 until streamItems.length()) {
                val name = streamItems.opt(index) as? String ?: return null
                val table = binaryByName[name] ?: continue
                if (!laneStreams.add(table)) return null
            }
            if (laneStreams.isEmpty()) return null
            return PushObjectLane(endpoint, maxObjectBytes, urlTtlSec, laneStreams)
        }

        private fun jsonInt64(value: Any?): Long? {
            if (value == null || value == JSONObject.NULL || value is Boolean) return null
            val number = value as? Number ?: return null
            val double = number.toDouble()
            if (double != double.toLong().toDouble()) return null
            return number.toLong()
        }

        private fun isCanonicalUuid(value: String): Boolean = runCatching {
            java.util.UUID.fromString(value).toString() == value
        }.getOrDefault(false)
    }
}

internal class PushConnectionTester(
    private val transportFactory: (PushEndpointPolicy.ValidEndpoint, String) -> PushTransport =
        { endpoint, token -> PushHttpTransport(endpoint, token) },
) {
    suspend fun test(
        endpoint: PushEndpointPolicy.ValidEndpoint,
        token: String,
    ): PushCapabilitiesResult = try {
        transportFactory(endpoint, token).capabilities()
    } catch (cancelled: kotlinx.coroutines.CancellationException) {
        throw cancelled
    } catch (_: Throwable) {
        val failure = PushFailure(PushFailureCode.NETWORK_IO)
        PushCapabilitiesResult.Rejected(failure.safeCode, failure.retryable, failure)
    }
}

internal fun canStartPushConnectionTest(
    networkAvailable: Boolean,
    endpointValid: Boolean,
    tokenAvailable: Boolean,
): Boolean = networkAvailable && endpointValid && tokenAvailable

sealed interface PushCapabilitiesResult {
    data class Available(val capabilities: PushCapabilities) : PushCapabilitiesResult
    data class Rejected(
        val reason: String,
        val retryable: Boolean,
        val failure: PushFailure? = null,
    ) : PushCapabilitiesResult
}
